Material FAB with speed dial implementation

AndyScherzinger opened the pull request:
This PR is for discussion and experimentally test driving a FAB implementation.

Nothing to see yet but I wanted to open a PR in order to check for build compliance and discuss the progress and changes/adaption of any feedback. Incorporating the discussion on #983

The branch is based on latest ````master```` as in release 1.9.1

<!---
@huboard:{"order":2.66015625,"milestone_order":1100}
-->

AndyScherzinger commented:
So here we go... Initial implementation of the Speed Dial FAB :) pinging @tobiasKaminsky @davivel and @jancborchardt 

The implementation work for ICS and up, is implemented as a Speed Dial (right now with labels since we do not have good icons to distinguish upload and upload from apps; they can just be taken out in the layout xml :smiley: ). It is linked already thus the actions work.

Open Issues:
* Action Bar icons left for now (since I don't now when they are still needed since the FAB might not be displayed)
* The FAB is hovering even in expanded mode but not blocking any other elements, thus you can expand the FAB Speed Dial and still click on list items (unsure if this is an expected behavior, or not though)
* ANT build is broken since the classpath doesn't get build up correctly atm (unsure why, as always works fine with gradle)... not saying supporting ant builds is frustrating :wink: Any idea what to do here @davivel ?

And here are the screenshots...

Lollipop
![device-2015-08-18-001613](https://cloud.githubusercontent.com/assets/1315170/9319053/6e0496b2-454a-11e5-8584-70583c106547.png)
![device-2015-08-18-001624](https://cloud.githubusercontent.com/assets/1315170/9319054/74375cae-454a-11e5-8d5d-dc7be3e2f4e8.png)

ICS
![device-2015-08-18-001749](https://cloud.githubusercontent.com/assets/1315170/9319055/78ff3770-454a-11e5-8669-3c9ed8102e73.png)
![device-2015-08-18-001859](https://cloud.githubusercontent.com/assets/1315170/9319057/7c082c06-454a-11e5-92cf-0f0d2e7dfd98.png)

AndyScherzinger commented:
...nevermind, found the classpath problem :tada:  :tada:  :tada:

AndyScherzinger commented:
@jancborchardt can you provide an icon for the "upload from app" button. I looked through the Google provided icons but couldn't find one that would have expressed the action at all :(

jancborchardt commented:
Hmmm … maybe https://www.google.com/design/icons/#ic_content_copy or https://www.google.com/design/icons/#ic_add ?

But I’d also be fine with having both be the same icons.

tobiasKaminsky commented:
Both actions the same icon is not so good as I do not like the text and thought, that the text will be removed.
If there is no suitable icon, maybe just open a choice dialog as it is now? And only have a "+" icon for both actions.
Although I would really like to have two meaningful icons...

AndyScherzinger commented:
I agree with @tobiasKaminsky that using one icon for two to be distinguished actions is rather confusing for the user. I put in the labels for now to be able to distinguish the actions but would also not want to go live with the labels.

As for the choice dialog that seems not quite right but probably a pragmatic solution for now and imho we should really aim for two fitting icons since that would lead to a proper use of the speed dial and wouldn't increase the number of click (still two as of today) instead of three if we would use a choice dialog.

The suggested icons don't seem to find in my opinion since the copy icon is used in a context where the user copies a selection on the same screen which wouldn't be the case here and thus an unexpected behavior, as for the plus icon that is already used as the icon om the FAB and thus shouldn't have two different actions triggered within the same context. I searched through https://materialdesignicons.com/ but couldn't find an appropriate icon for the app-upload. If someone has a nice icon which is related to "app" icon can create a derivation for app-upload :)

joeplus commented:
What about [this](https://materialdesignicons.com/icon/exit-to-app) one, rotated by 90° to left so that the arrow points to the top?

AndyScherzinger commented:
@jancborchardt @tobiasKaminsky what do you think? Not too bad, eyh?

joeplus commented:
Or flip the arrow as well like this:

![upload-from-app](https://cloud.githubusercontent.com/assets/7793652/9411685/b9f6623c-4827-11e5-83d8-7808fdd41c47.png)

AndyScherzinger commented:
I kind of prefer the recessed one but would shorten the arrow at the bottom be be more in balance

tobiasKaminsky commented:
I have found this:
![import](https://cloud.githubusercontent.com/assets/5836855/9429319/541203da-49cc-11e5-81d9-b058216e4a00.png)
https://materialdesignicons.com/icon/import

But I would like to have a generic app icon and a sending/import arrow.
I think that the user then would better understand it.

And when the text is gone, the text should appear if the user long presses the button. So the user can figure out what the button means instead of just clicking it.

AndyScherzinger commented:
Alternatively like a learning UI we could show the labels like in the screenshot for the first 1-3 times and after that don't show them anymore implying the user now know the function behind each mini fab.

what do you think?

tobiasKaminsky commented:
Learning UI is great, but still the text should appear when long-pressing (like in k-9)

AndyScherzinger commented:
@jancborchardt @tobiasKaminsky @joeplus any decision on the icons to be used?

@tobiasKaminsky where should the long click be implemented? on the mini fabs? And should that still be implemented with distinguishable icons?

AndyScherzinger commented:
Just added the implementation for the click events and cleaned up the action bar menu which now shows the remaining to actions :)

* click on the min fab gives you the according action
* long click on the mini fab shows a Toast naming the action behind the mini fab

--> for now I moved to the import icon hoping to jump start the icon discussion and getting to a point where we decide on the icons :)

![device-2015-09-11-150252](https://cloud.githubusercontent.com/assets/1315170/9815453/8a1a0f1c-5896-11e5-943e-b1df07c6f255.png)

jancborchardt commented:
Yeah, good! I guess at least for the first time opening the menu, the titles of the icons should be shown. Is that possible?

AndyScherzinger commented:
@jancborchardt absolutely, that is an easy one basically adding a preference value (under the hood, not in the preference screen) to store the fact, that the labels have been shown one time and if the preference has the value "already shown" the labels won't be added to the UI :) - easy.

That was what I meant with the learning UI.

AndyScherzinger commented:
any wishes for the coloring of the labels? I can change it to any colors you like :+1:

jancborchardt commented:
Cool! Yeah, just go with a default half-transparent grey or so, like what you had in a few screenshots above.

AndyScherzinger commented:
Done! 

It'll show labels until the user click on one of the smaller floating buttons and for the first session of the application.
The reason is that I couldn't find a way to remove the labels and resize the buttons click area, thus a transparent space stays until the next start of the application (start, not a pause/rusume!).

So whenever the list fragment gets initialized it checks if a fab ha ever been clicked before und if so doesn't add any labels.

AndyScherzinger commented:
How should we progress here?

Two questions are open imho, directing them to @jancborchardt:
* should we use the icon chosen for now?
* is the behavior regarding the showing/hiding of the labels okay?

If the icon and the behavior are okay I can finalize the PR adding the icon work in all the necessary resolutions, so maybe we can than have the two open material design PRs in 1.8.1 :grinning:

jancborchardt commented:
@AndyScherzinger yeah, I’m fine with the icon and the labels showing on first use. Thought my previous comment was already confirmation for that. :)

AndyScherzinger commented:
Cool! Well, the first use is not very specific so I asked to get a _very_ specific answer, so for now for the labels, first use basically means first session (in a business manor) and first activity life cycle (technically speaking)

AndyScherzinger commented:
@davivel @MTRichards @masensio @tobiasKaminsky @rperezb Ending the final discussions and details with @jancborchardt I consider this PR/Branch feature complete and ready for code review. So this would be the implementation to resolve #983 

Two important things (and I know I am repeating myself here):
* I only tested this PR on Lollipop (on a Nexus 5, 5.1.1)
* This PR is a child of #1090 (so 1090 would need to be merged first, or is deprecated if this one is merged first)

I don't want to put any pressure on anyone and I am fully aware of the release 1.8.0 being the top priority (kudos for that!) but just wanted to let you know that the implementation is done and ready for code review.

tobiasKaminsky commented:
I think we should do the review as soon as the previous commit #1090 is merged as otherwise this is too much to review.
But it is working fine for me (emulator).
Great job!

AndyScherzinger commented:
Thanks! :) I second that!

AndyScherzinger commented:
merged #1177

AndyScherzinger commented:
merged #1182

AndyScherzinger commented:
merged #1200 / #1192

AndyScherzinger commented:
jfyi open issue I found:
* atm the uploader screen also shows the fab (will investigate)

AndyScherzinger commented:
The FAB is part of the ExtendedListFragment implementation (its layout xml). Can anybody tell me when the FAB should be displayed as in on which subclasses and when it probably shouldn't? @davivel @masensio

tobiasKaminsky commented:
For me it is working as expected: Only the view that shows the content of the remote OC server is showing the FAB (class OCFileListListFragment).
So :+1: :+1: :+1:

AndyScherzinger commented:
The Button also shows up for copying/moving a file :(

tobiasKaminsky commented:
Ah. I missed this.
To answer your previous question: I guess the FAB should only be shown in OCFileListListFragment.

AndyScherzinger commented:
Sounds good to me :)
Would you agree with Tobias' point of view? @jancborchardt @davivel @masensio simply because his pov would be kind of done except for the FAB hiding on the mentioned screens anything else would need further implementation details.

jancborchardt commented:
What does that exactly mean in plain English? :D If you mean »Only show the big add button in the main Files view«, then yeah, I’d agree. ;)

AndyScherzinger commented:
That is what Tobias meant - at least that is what I understood :grin: 
Please correct me if I am wrong @tobiasKaminsky

AndyScherzinger commented:
This PR is now ready for reviews :) - @all please review (and merge) #1209 first

@LukeOwncloud @tobiasKaminsky can you do a community review? :)

jancborchardt commented:
Just thinking out loud here: Is it possible to display the icons around the fab in a circular manner instead of going up? Because now it’s a long way from the fab to the upload icon for example.

LukeOwncloud commented:
I noticed that the descriptive texts were gone when I wanted to really use the FAB for the first time. Is it intended that they do not appear after some time? Can they be brought back (without reinstalling the app)?

AndyScherzinger commented:
@LukeOwncloud that hiding after the first click on one of the mini fabs is designed by Jan. To get the description back you can long press the mini fabs :) 

@jancborchardt I can check but I am not sure the lib supports such a alignment and animation, besides that the top down alignment is what material design showed at some point, not sure though.

cornzy commented:
Just found this issue and as long it seems to be open I have another idea for the upload-from-app icon. 
The Play-(App-)Store Icon is just like a briefcase that's why I would prefere the briefcase-upload icon. The exit-to-app icon does not realy intuitivly leads to any upload action.

AndyScherzinger commented:
Thanks for your input @cornzy - I leave Design decisions with @jancborchardt - any comment Jan? :)

jancborchardt commented:
@cornzy I’d say let’s do that in a separate issue so we can finally get this over with. :) I’m happy to discuss details in another issue, best open one and supply screenshots.

AndyScherzinger commented:
@davivel, @masensio, @tobiasKaminsky and @przybylski calling for some git help :) This branch is based on ```material_buttons``` which already has the conflicts resolved against master and I tried to do another pull from material_buttons to materila_fab branch. It always tells me the branch is already up to date so I cannot pull in the changes (which is odd) can you guys try to update the branch for me (since I seems to not being able to get it done :disappointed: )

przybylski commented:
I hope I haven't broke anything

przybylski commented:
ok I broke like a lot, but thats weird it compiled on my machine :/

AndyScherzinger commented:
WOW! How did youo do this? It is building again. Oddly it doesn't show the latest changes from the ```material_buttons``` branch (like https://github.com/owncloud/android/commit/dd8c65dcd943dd9a0982baa978a41ddab6e48503 and e.g. https://github.com/owncloud/android/commit/a812784ddd925c89ac9b09f0481905d3541b2c47) but is says no conflicts with base branch ?! (I still don't understand git I guess...)

przybylski commented:
git branch -D material_buttons
git pull --all
git checkout --track origin/material_buttons

AndyScherzinger commented:
I mean pull the changed from material_buttons to material_fab ... It always tells me up-to-date when it obviously isn't :(

przybylski commented:
Those two changes are on branch (code is correct), maybe they are empty since material_buttons is rebased to master.

Hmm... thats odd, try delete both branches pull all and checkout them again

AndyScherzinger commented:
Hm, the changes are in the code but I would have expected the log to show the log items from the material_buttons branch then (which I can't see though). So I guess everything is fine then.

Again, thank you very much! :+1: I appreciate your help! :+1:

masensio commented on build.gradle:
```diff
@@ -20,10 +20,12 @@ repositories {
     compile 'com.android.support:support-v4:22.2.1'
     compile 'com.jakewharton:disklrucache:2.0.2'
     compile 'com.android.support:appcompat-v7:22.2.1'
+    compile 'com.getbase:floatingactionbutton:1.10.1'
```
I think it is better to use the Android Design Support Library to add de Floating Action Button, in other to not  add more third-party dependecies.

AndyScherzinger replied:
The problem with the design support lib is that it does not support the speed dial but just simple FABs. That is why I added the third party lib. Else we would have to write our own implementation.

AndyScherzinger replied:
.. So we could for example also move on to this lib: https://github.com/yavski/fab-speed-dial which needs the 23.1.1 lib (might be back-portable if even necessary) because it implements the speed dial with the Support Lib FAB implementation. But as I said before we would need to use some lib for the speed dial or someone would have to implement it within this app.

So how about I integrate the needed lilb classes into our third party code folder? Would that be feasable?

davivel replied:
Speed dial, understood. That's a good reason to keep the dependencies as they are right now.

The point of using FAB from Support Lib was reducing the number of dependency providers, but should not be with the cost of losing functionality.

I wouldn't the classes in 'third_party', keep the dependency in Gradle should be prefered when possible.

AndyScherzinger replied:
Sweet, tan we can keep the lib for now. Whenever a speeddial implementation would be available within googles support lib like "design" I would than refactor it and remove the third party lib... continuous improvement so to say :smiley:

davivel commented:
@DeepDiver1975 , cla-bot-android is failing here due to excess of requests to Github (``` **429 Too Many Requests** ```).

Any way to throw the check again without sending fake commits?

AndyScherzinger commented:
@davivel anything that should be done to push this PR forward? ..except me updating to the latest master and resolving the conflicts :smiley:

rperezb commented:
@AndyScherzinger this will be pushed during the next 2 weeks, we have just priorizated it with @MTRichards to be qa and merge

davivel commented:
And with next 2 weeks @rperezb means let's start with it now :D 

If you could solve conflicts would be awesome.

AndyScherzinger commented:
Hi @davivel I tried my best but I still don't understand git and rebase. SVN4Life! ;)
It think I:
* rebased to master
* fixed all conflicts
* test drove the branch and it works on my Nexus5X

The build process seems to be hanging and doesn't give me any log output to check why that is...
But for some reason my rebase simply doubled the number of commits within this PR shown at the top which doesn't make any sense to me :cry: 
Still hope it helps and also works for you guys.

jancborchardt commented:
Hey @AndyScherzinger I'll be in FFM today starting 10:00! Let me know if you're up for coworking or lunch! :) My email is jan at owncloud dot com

masensio commented:
Hi @AndyScherzinger,

thanks for updating the branch. I've seen your commits are duplicated (as you said). Maybe your last merge is the reason of it.
I think the code works for testing. 
I don't know if it better to squash this code before testing, to avoid this duplicated commits. At the moment it is a crazy PR with duplicated commits. @davivel what do you think?

AndyScherzinger commented:
Hi @masensio - yes please! 

It took me forever to even get the rebase done/working since I never did this and I probably did a lot of things wrong rebasing the branch, so my issue is that even if I rebase again due to my lack of knowledge I most likely will make it even worse. :disappointed: 
I got so frustrated when rebasing that I honestly though about reimplementing the whole thing again with a new master based branch since the number of changes except for the lib artefacts isn't huge.

AndyScherzinger commented:
Only the stuff below @davivel's comment https://github.com/owncloud/android/pull/1100#issuecomment-188869134 is imho relevant for the fab implementation

masensio commented:
@AndyScherzinger, don't worry about the rebase! I'll help you

As first step, undone the latest commit, and check the log. Maybe more than this commit have dissappear because that merge include other commits.
Maybe you need to remove other commits more, as I see in the list of commits.
This first step is for having the status before your rebase.

Then we need to rebase again with master. For this step:
- update your master branch (`git rebase origin/master`)
- `git checkout material_fab`
- `git rebase master`

Here you can have conflicts, you need to solve them and save and continue with the rebase
- `git rebase --continue`

And you need to continue fixing conflict and rebase --continue until the end of the rebase

When you will fix everything: `git push -f origin material_fab`

NOTE: origin is your remote repo.

I think here you have the instructions to do the rebase, you can do the first step and I'll help you with the rest of process tomorrow if it is neccessary. :four_leaf_clover:

AndyScherzinger commented:
Hi @masensio 
Thanks for the help and info. Doesn't work :( (my fault!). I did everything according to your how-to but afert alls with the last command (git push -f origin material_fab) I just get a "everything-up-to-date" and nothing happens. The conflicts to be resolved a rather a joke (5 files, all related to the build process)

AndyScherzinger commented:
Hi @masensio 
my mistake, the merges are killing me this is impossible, I am sorry. I will rather completely reimplement the fab based on the actual branch and the done changes than trying to merge and resolve these conflicts. After 6 month this is just impossible and frustrating... I am sorry :disappointed:

masensio commented:
Hi @AndyScherzinger,

keep calm and wait a moment. I'll try to undone the merge you have done and then we can rebase again.
If this try fails I'll tell you.

AndyScherzinger commented:
Hi @masensio 
no worries, I also didn't mean it in an offensive way. I am not the most patient person and couldn't get it working quickly with bothered me :( My mistake! But like I said, I can also re-implement it checking all the diffs of my commits. They are not that many changes that need to be made, most of my commits where related to getting the lib properly integrated with the ant build. Again I hope I didn't offend you or anyone from the core team, if so it wasn't my intention.

AndyScherzinger commented:
@masensio you are awesome :tada: :gem: I tested the latest branch on my Nexus5X and everything is working for me! This is incredible!!!! This truely made my day! :smiley: 

@jesmrec what I tested:
* device rotation with fab opened
* back button with fab and drawer opened
* labels hidden after initial display and app restart
*  toast displaying the action on long click on the mini fab(s)
* fab collapsed after click on mini fab(s)

jesmrec commented:
@AndyScherzinger thank you!! i'm pretty sure i'll work fine.

rperezb commented:
@AndyScherzinger great! you may check the test plan the @jesmrec has designed on: https://github.com/owncloud/QA/pull/161 (feeedback welcome)

davivel commented:
Hi, @AndyScherzinger . It's totally normal you get frustrated with rebases. We have been handling git more often and still suffer that from time to time.

And don't worry, nobody here was offended for nothing you said. Tons of hugs for you :-) 

For the future, just consider trying ```git rebase master``` instead of ```git merge master```when bringing changes from master to a branch where you are working. That way the history of the branch is easier to read. It's much easier if you try this from the very beginning in a fresh branch, than trying to fix it later, when there are tons of commits. 

Maybe this helps to you: https://git-scm.com/book/en/v2/Git-Branching-Rebasing . Understanding what happens with the history with the different git commands made my life a lot easier when struggling with the wonders of rebasing :)

davivel commented on libs/com-getbase-floatingactionbutton-1-10-1-exploded-aar/AndroidManifest.xml:
```diff
@@ -0,0 +1,14 @@
+
+    <uses-sdk
+        android:minSdkVersion="14"
+        android:targetSdkVersion="22" />
```
This should be updated to target 23, if possible

AndyScherzinger replied:
Yes, exactly. That is why I left it untouched. Still hope Google will provide a speed dial fab at some point... :)

davivel commented on res/layout/list_footer.xml:
```diff
@@ -1,7 +1,7 @@
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="match_parent"
-    android:layout_height="56dp"
+    android:layout_height="112dp"
```
This is way bigger. What was the problem? The FAB hiding the message?

AndyScherzinger replied:
It is done so that the List scrolls beyond the fab else it would hover above the last list item and for example that  fact that the last item is shared via link could not be seen in the list, see screenshot with the 1121dp height:
![device-2016-03-02-112556](https://cloud.githubusercontent.com/assets/1315170/13457661/90464b0c-e069-11e5-8b7b-0b5f1d5712c9.png)

With the height of 112 the fab is perfectly aligned in the footer with 16dp to the right, top and bottom :smiley:

davivel replied:
:+1:

davivel commented on res/layout/list_footer.xml:
```diff
@@ -14,7 +14,8 @@
         android:layout_gravity="center"
         android:gravity="center"
-        android:textColor="@color/setup_text_hint"
+        android:textColor="@color/secondaryTextColor"
```
I - love - YOU

AndyScherzinger replied:
:heart_eyes: happy to help!

davivel commented on res/layout/list_fragment.xml:
```diff
@@ -81,4 +85,52 @@
+        android:layout_alignParentBottom="true"
+        android:layout_alignParentRight="true"
+        android:layout_alignParentEnd="true"
+        fab:fab_addButtonColorNormal="@color/owncloud_blue_accent"
```
This probably will need to be changed to ```primary_button_background_color```. But let us check it, maybe we need to refactor the colors available in setup.xml

davivel replied:
Or still better: to ```color_accent```

AndyScherzinger replied:
Yes you are right, ```color_accent``` would be the right one, unfortunately not part of the ```setup.xml``` at the moment so it couldn't be customized :(
I'll change it to color_accent, open issue then remains to make this available for branding. --> https://github.com/owncloud/android/commit/6c83af4521c5566d353e3ac281689679ad35b24f

using primary_button_background_color would enable branding though, but wouldn't be the right "name". So we could use it so branding still works for the next release plus I would really like to rewrite setup and color XMLs to have appropriate color definitions in setup.xml and then to the internally used names in the color.xml thus everything could be branded easily.

If this is in line with your strategy I would open a new issue where we can have a discussion on how to proceed with this and I can get a better understanding about how the setup.xml is used and after that rework the color definition :)

davivel replied:
We need to study it internally , don't worry too much about this right now, I will come back with something more concrete.

AndyScherzinger commented:
Yeah, I tried the rebase with the drawer branch yesterday evening... duplicated all master commits... 
I hopefully make it to the conference in September so someone can teach me this (I am only using tortoiseGit, so no command line tools), because I don't seem to get it...

davivel commented on setup_env.bat:
```diff
@@ -24,6 +24,7 @@ call git submodule update
 
 call android.bat update project -p libs/android-support-appcompat-v7-exploded-aar --target android-23
 call android.bat update project -p libs/android-support-design-exploded-aar --target android-23
+call android.bat update project -p libs/com-getbase-floatingactionbutton-1-10-1-exploded-aar --target android-23
```
The target number here should coincide with the value of targetSdk in libs/com-getbase-floatingactionbutton-1-10-1-exploded-aar/AndroidManifest.xml ; would be awesome if 23 works, though if 22 was the original in the Manifest, problems could arise.

AndyScherzinger replied:
DONE, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61 - kept v22 since it doesn't make a difference anyways since the lib is in binary form thus compiled against v22 anyways... :D

davivel commented on setup_env.sh:
```diff
@@ -19,6 +19,7 @@ function initForAnt {
 	#Prepare project android-support-appcompat-v7 ; JAR file is not enough, includes resources
 	android update lib-project -p libs/android-support-appcompat-v7-exploded-aar --target android-23
```
Same here

AndyScherzinger replied:
DONE, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61

davivel commented on src/com/owncloud/android/ui/activity/FileDisplayActivity.java:
```diff
@@ -144,7 +144,7 @@
 
     private static String DIALOG_UNTRUSTED_CERT = "DIALOG_UNTRUSTED_CERT";
-    private static String DIALOG_CREATE_FOLDER = "DIALOG_CREATE_FOLDER";
+    public static String DIALOG_CREATE_FOLDER = "DIALOG_CREATE_FOLDER";
```
Really needs to be ```public```? I don't find where. Would prefer ```private```, if possible.

davivel replied:
Testing a bit, and seems working fine.

AndyScherzinger replied:
It is used in FileDisplayActivity and in OCFileListFragment

davivel replied:
You also created public ```FileDisplayActivity#createFolder() ```; could ```OCFileListFragment``` call it instead of directly creating a dialog? It's duplicated code. That way the constant may stay private also.

AndyScherzinger replied:
I slightly refactored the code.
* the dialog isn't used in the Activity anymore
  - so I removed the method in the activity https://github.com/owncloud/android/commit/8d038e89a9560e190f282d5b074b909193292822
  - moved the requestCode to the fragment https://github.com/owncloud/android/commit/f34a9ddf5b2c2a5bb326fc08ddac65c3e3546737
  - also removed the "other" action implementations since they are now handeled by the fab implementation in the fragment https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac
* shortened the upload call by creating a static helper method within the upload activity https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac

davivel commented on src/com/owncloud/android/ui/activity/FileDisplayActivity.java:
```diff
@@ -863,8 +896,6 @@ public void onBackPressed() {
                 setFile(listOfFiles.getCurrentFile());
             }
             cleanSecondFragment();
-        } else {
```
Why removed this code? Should be there to close the app, shouldn't it?

davivel replied:
Testing a bit, and seems working fine. (sorry, wrong comment before)

AndyScherzinger replied:
See line 882, the final else has moved :)

davivel replied:
:+1:

davivel commented on src/com/owncloud/android/ui/activity/FileDisplayActivity.java:
```diff
@@ -943,6 +974,14 @@ protected void onPause() {
     }
 
+    public boolean isFabOpen() {
+        if(getListOfFilesFragment() != null && getListOfFilesFragment().getFabMain() != null && getListOfFilesFragment().getFabMain().isExpanded()) {
```
Condition too long, please, split in two lines.

AndyScherzinger replied:
DONE, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61

davivel commented on src/com/owncloud/android/ui/fragment/ExtendedListFragment.java:
```diff
@@ -64,6 +69,11 @@
     private SwipeRefreshLayout mRefreshEmptyLayout;
     private TextView mEmptyListMessage;
+
+    private FloatingActionsMenu fabMain;
```
Please, let's use the ```m``` prefix for member variables. That helps to recognize members in code snippets in Github, where syntax highlight is not applied to them.

AndyScherzinger replied:
DONE, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61

davivel commented on src/com/owncloud/android/ui/fragment/ExtendedListFragment.java:
```diff
@@ -327,6 +357,23 @@ public void setSwipeEnabled(boolean enabled) {
     }
 
     /**
+     * Disables FAB.
```
Seems this line is not needed.

AndyScherzinger replied:
Why? It is part of the java doc for the disable fab method.

davivel replied:
The method allows both disabling or enabling. Removing the line the rest of the comment seemed to describe perfectly the method. I actually thought this was a forgotten line of a previous version of the method.

Of course, this is unimportant, sorry  for being so punctilious :$

AndyScherzinger replied:
No worries, I like your eyes for the details! :+1: 
Done, see https://github.com/owncloud/android/commit/4327cf5433744611248630b48c49dbcee04048e3

davivel commented on src/com/owncloud/android/ui/fragment/OCFileListFragment.java:
```diff
@@ -154,8 +205,167 @@ public void onActivityCreated(Bundle savedInstanceState) {
+        getFabUploadFromApp().setTitle(null);
+        ((TextView) getFabUpload().getTag(com.getbase.floatingactionbutton.R.id.fab_label)).setVisibility(View.GONE);
+        ((TextView) getFabMkdir().getTag(com.getbase.floatingactionbutton.R.id.fab_label)).setVisibility(View.GONE);
+        ((TextView) getFabUploadFromApp().getTag(com.getbase.floatingactionbutton.R.id.fab_label)).setVisibility(View.GONE);
```
Line too long. Plase, splint in two.

AndyScherzinger replied:
DONE, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61

davivel commented on src/com/owncloud/android/ui/fragment/OCFileListFragment.java:
```diff
@@ -118,7 +132,44 @@ public void onAttach(Activity activity) {
+        Log_OC.i(TAG, "onCreateView() start");
+        View v = super.onCreateView(inflater, container, savedInstanceState);
+
+        // Setup FAB listeners
```
The setup of FAB listeners is being done twice, with some slight differences: one here and another in ```onActivityCreatedView()```.

One of both should be removed, probably this one.

AndyScherzinger replied:
DONE - onActivityCreated is the right one, see https://github.com/owncloud/android/commit/e4d936e83754b823c90160889ce4704dfc3bbe61

davivel commented:
Code reviewed. Really good job here, very minor things to update.

Almost ready!

AndyScherzinger commented:
@davivel Thanks for the code review, changes have been commited or I added comments for further clarification :+1:

davivel commented on src/com/owncloud/android/ui/fragment/OCFileListFragment.java:
```diff
@@ -154,8 +177,170 @@ public void onActivityCreated(Bundle savedInstanceState) {
+                );
+                //startActivityForResult(action, ACTION_SELECT_MULTIPLE_FILES);
+                // this flow seems broken;
+                getActivity().startActivityForResult(
```
```FileDiplayActivity``` also exposes a public method to do this , would prefer it's used here.

AndyScherzinger replied:
As mentioned in the other comment, this is now exclusively handled by the fragment, since the activity doesn't trigger these actions anymore, I slightly refactored the code
* so I removed the method in the activity https://github.com/owncloud/android/commit/8d038e89a9560e190f282d5b074b909193292822
* moved the requestCode to the fragment https://github.com/owncloud/android/commit/f34a9ddf5b2c2a5bb326fc08ddac65c3e3546737
* also removed the "other" action implementations since they are now handeled by the fab implementation in the fragment https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac
* shortened the upload call by creating a static helper method within the upload activity https://github.com/owncloud/android/commit/7e82baa8bc068bee956f6f03b61c06151c953eac

davivel commented:
Thanks, @AndyScherzinger . Answered again.

AndyScherzinger commented:
Thanks for the thorough code review, I really think this pushed the PRs quality :+1: 
Please check my comments since I went for a slightly different solution in comparison to your comments. Hope this is fine with you :smiley:

AndyScherzinger commented:
@davivel forgot to mention that if there is anything else regarding the PR and its implementatation please let me know! :smiley:

davivel commented:
I like what you did :) .

This is :+1: from dev POVs, time for testing. 

Thank you so much, pal.

davivel commented:
@AndyScherzinger , maybe you want to have a look to the test plan in https://github.com/owncloud/QA/pull/161 . If find anything wrong, or miss something, feel free to comment :)

AndyScherzinger commented:
Glad I can contribute to owncloud and again thank you very much for your input and the time you spend on my PRs. I am aware that you have a lot of tasks in the oC ecosystem, customer specific tasks, etc. and I am thankful that you still make time to support community contributions!

I'll check the QA task, pleasure to help!

davivel commented:
@AndyScherzinger , I modified the color set for normal state of FAB so that it's the same as for 'primary buttons'.

Consider this a temporary fix to complete test pass. We will review the needs of customizable colours soon, and hopefully define clear criteria to apply them in the future. I will keep everybody, but specially you, informed about it.

cc @jesmrec , ready for final test.

AndyScherzinger commented:
@davivel yeah, perfectly fine with me :) :+1: 
@jesmrec woohoo! QA approved! This is awesome! :+1:

davivel commented:
Then everything is OK. Pa' la saca. :)

AndyScherzinger commented:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: 
*WOOHOO*
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

supportreq commented:
@AndyScherzinger one query, any particular reason for the FAB labels to be disabled post first restart of the app? I got confused as why the label is missing after restart.. can we have it as a setup.xml preference..

AndyScherzinger commented:
sure thing, that would then have to be discussed with @davivel and @jancborchardt 
In general you can still long-press the mini fabs which will give you the label as a toast.
